### ISSUE TITLE
Invalid argument type on generic constructor call can leak typevar instead of solving to Unknown

### ISSUE THREAD

**Issue by gpollo** [astral-sh/ty#3944]
### Summary

Hi, I've encountered the following error while writing some code recently.

Sorry if it's duplicated. I couldn't properly describe the problem while searching existing bugs. I've removed most of the code to get a small MRE. Don't worry if it doesn't make too much sense.

```python
from __future__ import annotations

from dataclasses import dataclass
from typing import Callable, Protocol, TypeVar, Generic, Final

T = TypeVar("T", contravariant=True)

class ModbusMapper(Protocol[T]):
    def into_modbus(self, value: T) -> int: ...

@dataclass(frozen=True)
class ModbusMapperFuncs(ModbusMapper[T]):
    func_into_modbus: Callable[[T], int] | None = None

    def into_modbus(self, value: T) -> int:
        if self.func_into_modbus is None:
            raise RuntimeError("missing conversion function")
        return self.func_into_modbus(value)

class MappedRegister(Generic[T]):
    def __init__(self, callbacks: ModbusMapper[T]) -> None:
        self.callbacks: Final[ModbusMapper[T]] = callbacks

class IntegerRegister:
    def map(
        self, mapper: ModbusMapper[T] | Callable[[T], int]
    ) -> MappedRegister[T]:
        callbacks: ModbusMapper[T]

        if callable(mapper):
            # ERROR: This should be valid, since ModbusMapperFuncs[T] implements ModbusMapper[T].
            callbacks = ModbusMapperFuncs(func_into_modbus=mapper)
        else:
            callbacks = mapper

        return MappedRegister(callbacks)
```

Running `uv run ty check test.py` returns the following error.

```
error[invalid-assignment]: Object of type `ModbusMapperFuncs[T@ModbusMapperFuncs]` is not assignable to `ModbusMapper[T@map]`
  --> test.py:37:13
   |
37 |             callbacks = ModbusMapperFuncs(func_into_modbus=mapper)
   |             ---------   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Incompatible value of type `ModbusMapperFuncs[T@ModbusMapperFuncs]`
   |             |
   |             Declared type `ModbusMapper[T@map]`
   |
info: type `ModbusMapperFuncs[T@ModbusMapperFuncs]` is not assignable to protocol `ModbusMapper[T@map]`
info: └── protocol member `into_modbus` is incompatible
info:     └── parameter `value` has an incompatible type: `T@map` is not assignable to `T@ModbusMapperFuncs`
```

It thinks that

* `ModbusMapperFuncs[T@ModbusMapperFuncs]`

isn't compatible with

* `ModbusMapper[T@map]`.

But, `T@ModbusMapperFuncs` is `T@map` as expected as it comes from the same generic definition.

This code passes without errors in `mypy` and also Pylance.

### Version

ty 0.0.56 (3078f9fa2 2026-07-01)

**Comment by carljm** [comment]
Thanks for the report! Here's a simpler version (eliminating protocols, dataclasses, inheritance, and other irrelevant factors) that reproduces the same issue:

```py
from typing import Generic, TypeVar, Callable

T = TypeVar("T")
U = TypeVar("U")

class Box(Generic[U]):
    def __init__(self, func: Callable[[U], int]) -> None: ...

def test(value: Callable[[T], int] | int) -> None:
    if callable(value):
        target: Box[T] = Box(value)
```

The root cause here is #1578. This is technically correct, and is currently expected behavior in ty, though it differs from other type-checkers, and may be overly pedantic. That is, ty sees that the check `if callable(value):` does not actually guarantee that `value` is of type `Callable[[T], int]`, because a subclass of `int` can add a `__call__` method and become callable, and might have any signature. This causes an `invalid-argument-type` error (which you didn't include in your report, but is really the root cause here) trying to assign `value` to `Callable[[U], int]` in the constructor of `Box`.

The `invalid-assignment` is a cascading error, which looks like a bug. When we infer the type of `Box(value)`, due to the `invalid-argument-type` we get no information from the `value` argument. This leaves an implicit tautological `U@Box = U@Box` constraint (arising from the implicit `self` argument) as the sole constraint on `U`. We then "solve" `U` to itself, rather than falling back to `Unknown` as we should.

Focusing this issue on that cascading generic inference bug, since #1578 already covers the root cause of why we error here in the first place.

**Comment by gpollo** [comment]
Thank you for explaining the problem. I now think the current behavior is sound when explained that way.

I'll see if I can rework my API for type correctess, or else just ignore the error. Feel free to close the issue if you want. Thanks!
